warehouse/annotation-core/src/main/java/datawave/annotation/data/v1/AccumuloAnnotationSerializer.java

```diff
@@ -0,0 +1,235 @@
+import datawave.annotation.protobuf.v1.Segment;
+
+public class AccumuloAnnotationSerializer implements AnnotationSerializer<Iterator<Map.Entry<Key,Value>>,Annotation> {
+    public static final char NULL = 0x0;
+    public static final Value EMPTY = new Value();
+
+    protected static final Logger log = LoggerFactory.getLogger(AccumuloAnnotationSerializer.class);
+
+    final AnnotationVisibilityTransformer visibilityTransformer;
+
+    public AccumuloAnnotationSerializer() {
+        this(new DefaultAnnotationVisibilityTransformer());
+    }
+
+    public AccumuloAnnotationSerializer(AnnotationVisibilityTransformer visibilityTransformer) {
+        this.visibilityTransformer = visibilityTransformer;
+    }
+
+    @Override
+    public Iterator<Map.Entry<Key,Value>> serialize(Annotation annotation) throws AnnotationSerializationException {
+        // TODO: validate that the annotation has one or more segments
+        // TODO: validate that the annotation has a valid shard (at least?)
+        Key baseKey = generateBaseKey(annotation);
+        List<Map.Entry<Key,Value>> serializedResults = new ArrayList<>();
+
+        // TODO: convert the annotation id to a string/byte array once and pass this down instead of converting it each
+        // time
+        if (!annotation.getMetadataMap().isEmpty()) {
+            serializeMetadata(baseKey, annotation.getAnnotationId(), annotation.getMetadataMap(), serializedResults);
+        }
+
+        serializeSegments(baseKey, annotation.getAnnotationId(), annotation.getSegmentsList(), serializedResults);
+
+        return serializedResults.iterator();
+    }
+
+    @Override
+    public Annotation deserialize(Iterator<Map.Entry<Key,Value>> elements) throws AnnotationSerializationException {
+        if (elements == null || !elements.hasNext()) {
+            return null;
+        }
+
+        final Annotation.Builder annotationBuilder = Annotation.newBuilder();
+        final List<Segment> segments = new ArrayList<>();
+        final Map<String,String> metadata = new HashMap<>();
+        Key baseKey = null;
+
+        while (elements.hasNext()) {
+            Map.Entry<Key,Value> e = elements.next();
+            final Key key = e.getKey();
+
+            if (baseKey == null) {
+                baseKey = key;
+            } else if (!correctAnnotationSource(baseKey, key)) {
+                throw new AnnotationSerializationException("The key provided isn't from the same annotation as the " + "first key provided: baseKey: ["
+                                + baseKey + "] currentKey: [" + key + "]");
+            }
+
+            if (log.isTraceEnabled()) {
+                log.trace("Iterated key: '{}'", e.getKey());
+            }
+
+            String[] cfParts = key.getColumnFamily().toString().split("\0");
+            if (cfParts.length != 3) {
+                throw new AnnotationSerializationException("Column family in key didn't have 2-3 parts, key: '" + key + "'");
```

Comment:
```suggestion
                throw new AnnotationSerializationException("Column family in key didn't have 3 parts, key: '" + key + "'");
```

Reply:
at first I thought this was a typo, but given the code below it seems correct, and the check above may actually be wrong